// babylonjs-editor/build/electron/routes/storage.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const electron_1 = require("electron");
const fs = require("fs-extra");
const path = require("path");
const KoaRouter = require("koa-router");
class StorageRouter {
    /**
     * Constructor
     * @param application: the KOA application
     */
    constructor(application) {
        // Protected members
        this.path = '';
        this.application = application;
        this.router = new KoaRouter();
        // Create routes
        this.getFiles();
        this.setWorkingDirectory();
        this.writeFile();
        this.readFile();
        this.createFolder();
        this.getPaths();
        this.getNewFilePath();
        this.application.use(this.router.routes());
    }
    /**
     * Returns the files
     */
    getFiles() {
        this.router.get('/files', (ctx, next) => __awaiter(this, void 0, void 0, function* () {
            this.path = (ctx.query && ctx.query.path) ? ctx.query.path : process.cwd();
            const entries = yield fs.readdir(this.path);
            const files = [{ name: '..', folder: path.join(this.path, '..') }];
            for (const e of entries) {
                const stat = yield fs.stat(path.resolve(this.path, e));
                files.push({ name: e, folder: stat.isDirectory() ? path.resolve(this.path, e) : null });
            }
            ctx.body = {
                value: files
            };
        }));
    }
    /**
     * Set the working directory
     */
    setWorkingDirectory() {
        this.router.post('/files:/workingDirectory', (ctx, next) => __awaiter(this, void 0, void 0, function* () {
            this.path = ctx.request.body.name;
            ctx.body = {
                message: 'success'
            };
        }));
    }
    /**
     * Writes the given file
     */
    writeFile() {
        this.router.put('/files:/write', (ctx, next) => __awaiter(this, void 0, void 0, function* () {
            if (!ctx.query || !ctx.query.name || !ctx.query.folder)
                throw new Error('Please provide a folder to write file in.');
            const chunks = [];
            ctx.req.on('data', (chunk) => {
                chunks.push(chunk);
            });
            ctx.req.on('end', () => __awaiter(this, void 0, void 0, function* () {
                const buffer = Buffer.concat(chunks);
                // Write file
                yield fs.writeFile(path.join(ctx.query.folder, ctx.query.name), buffer);
            }));
            ctx.body = {
                message: 'success'
            };
        }));
    }
    /**
     * Get file content
     */
    readFile() {
        this.router.get('/files:/read', (ctx, next) => __awaiter(this, void 0, void 0, function* () {
            const filename = ctx.query.path;
            ctx.type = path.extname(filename);
            ctx.body = fs.createReadStream(filename, {
                encoding: 'utf-8'
            });
        }));
    }
    /**
     * Creates a folder
     */
    createFolder() {
        this.router.post('/files:/folder', (ctx, next) => __awaiter(this, void 0, void 0, function* () {
            const folder = ctx.request.body.name;
            const filename = path.resolve(this.path, folder);
            if (!(yield fs.pathExists(filename)))
                yield fs.mkdir(filename);
            ctx.body = {
                message: 'success'
            };
        }));
    }
    /**
     * Get a path
     */
    getPaths() {
        this.router.get('/files:/paths', (ctx, next) => __awaiter(this, void 0, void 0, function* () {
            const paths = yield new Promise((resolve, reject) => {
                electron_1.dialog.showOpenDialog({
                    title: ctx.query && ctx.query.title ? ctx.query.title : undefined,
                    defaultPath: ctx.query && ctx.query.folder ? ctx.query.folder : undefined,
                    properties: [ctx.query && ctx.query.type || 'openDirectory']
                }, paths => resolve(paths));
            });
            ctx.body = paths;
        }));
    }
    /**
     * Get a full file name
     */
    getNewFilePath() {
        this.router.get('/files:/newfilepath', (ctx, next) => __awaiter(this, void 0, void 0, function* () {
            const paths = yield new Promise((resolve, reject) => {
                electron_1.dialog.showOpenDialog({
                    title: ctx.query && ctx.query.title ? ctx.query.title : undefined,
                    defaultPath: ctx.query && ctx.query.folder ? ctx.query.folder : undefined,
                    properties: ['promptToCreate']
                }, paths => resolve(paths));
            });
            ctx.body = paths;
        }));
    }
}
exports.default = StorageRouter;
